{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Error Handling\n",
    "\n",
    "Implement various kinds of error handling and resource management.\n",
    "\n",
    "An important point of programming is how to handle errors and close\n",
    "resources even if errors occur.\n",
    "\n",
    "This exercise requires you to handle various errors. Because error handling\n",
    "is rather programming language specific you'll have to refer to the tests\n",
    "for your track to see what's exactly required.\n",
    "\n",
    "## Hints\n",
    "\n",
    "For the `filelike_objects_are_closed_on_exception` function, the `filelike_object`\n",
    "will be an instance of a custom `FileLike` class defined in the test suite. This\n",
    "class implements the following methods:\n",
    "- `open` and `close`, for explicit opening and closing.\n",
    "- `__enter__` and `__exit__`, for implicit opening and closing.\n",
    "- `do_something`, which may or may not throw an `Exception`.\n",
    "\n",
    "\n",
    "## Exception messages\n",
    "\n",
    "Sometimes it is necessary to raise an exception. When you do this, you should include a meaningful error message to\n",
    "indicate what the source of the error is. This makes your code more readable and helps significantly with debugging. Not\n",
    "every exercise will require you to raise an exception, but for those that do, the tests will only pass if you include\n",
    "a message.\n",
    "\n",
    "To raise a message with an exception, just write it as an argument to the exception type. For example, instead of\n",
    "`raise Exception`, you shold write:\n",
    "\n",
    "```python\n",
    "raise Exception(\"Meaningful message indicating the source of the error\")\n",
    "```\n",
    "\n",
    "## Submitting Exercises\n",
    "\n",
    "Note that, when trying to submit an exercise, make sure the solution is in the `$EXERCISM_WORKSPACE/python/error-handling` directory.\n",
    "\n",
    "You can find your Exercism workspace by running `exercism debug` and looking for the line that starts with `Workspace`.\n",
    "\n",
    "For more detailed information about running tests, code style and linting,\n",
    "please see the [help page](http://exercism.io/languages/python).\n",
    "\n",
    "## Submitting Incomplete Solutions\n",
    "It's possible to submit an incomplete solution so you can see how others have completed the exercise.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def handle_error_by_throwing_exception():\n",
    "    pass\n",
    "\n",
    "\n",
    "def handle_error_by_returning_none(input_data):\n",
    "    pass\n",
    "\n",
    "\n",
    "def handle_error_by_returning_tuple(input_data):\n",
    "    pass\n",
    "\n",
    "\n",
    "def filelike_objects_are_closed_on_exception(filelike_object):\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "\n",
    "\n",
    "\n",
    "class FileLike(object):\n",
    "    def __init__(self, fail_something=True):\n",
    "        self.is_open = False\n",
    "        self.was_open = False\n",
    "        self.did_something = False\n",
    "        self.fail_something = fail_something\n",
    "\n",
    "    def open(self):\n",
    "        self.was_open = False\n",
    "        self.is_open = True\n",
    "\n",
    "    def close(self):\n",
    "        self.is_open = False\n",
    "        self.was_open = True\n",
    "\n",
    "    def __enter__(self):\n",
    "        self.open()\n",
    "        return self\n",
    "\n",
    "    def __exit__(self, *args):\n",
    "        self.close()\n",
    "\n",
    "    def do_something(self):\n",
    "        self.did_something = True\n",
    "        if self.fail_something:\n",
    "            raise Exception(\"Failed while doing something\")\n",
    "\n",
    "\n",
    "class ErrorHandlingTest(unittest.TestCase):\n",
    "    def test_throw_exception(self):\n",
    "        with self.assertRaisesWithMessage(Exception):\n",
    "            er.handle_error_by_throwing_exception()\n",
    "\n",
    "    def test_return_none(self):\n",
    "        self.assertEqual(er.handle_error_by_returning_none('1'), 1,\n",
    "                         'Result of valid input should not be None')\n",
    "        self.assertIsNone(er.handle_error_by_returning_none('a'),\n",
    "                          'Result of invalid input should be None')\n",
    "\n",
    "    def test_return_tuple(self):\n",
    "        successful_result, result = er.handle_error_by_returning_tuple('1')\n",
    "        self.assertIs(successful_result, True,\n",
    "                      'Valid input should be successful')\n",
    "        self.assertEqual(result, 1, 'Result of valid input should not be None')\n",
    "\n",
    "        failure_result, result = er.handle_error_by_returning_tuple('a')\n",
    "        self.assertIs(failure_result, False,\n",
    "                      'Invalid input should not be successful')\n",
    "\n",
    "    def test_filelike_objects_are_closed_on_exception(self):\n",
    "        filelike_object = FileLike(fail_something=True)\n",
    "        with self.assertRaisesWithMessage(Exception):\n",
    "            er.filelike_objects_are_closed_on_exception(filelike_object)\n",
    "        self.assertIs(filelike_object.is_open, False,\n",
    "                      'filelike_object should be closed')\n",
    "        self.assertIs(filelike_object.was_open, True,\n",
    "                      'filelike_object should have been opened')\n",
    "        self.assertIs(filelike_object.did_something, True,\n",
    "                      'filelike_object should call do_something()')\n",
    "\n",
    "    def test_filelike_objects_are_closed_without_exception(self):\n",
    "        filelike_object = FileLike(fail_something=False)\n",
    "        er.filelike_objects_are_closed_on_exception(filelike_object)\n",
    "        self.assertIs(filelike_object.is_open, False,\n",
    "                      'filelike_object should be closed')\n",
    "        self.assertIs(filelike_object.was_open, True,\n",
    "                      'filelike_object should have been opened')\n",
    "        self.assertIs(filelike_object.did_something, True,\n",
    "                      'filelike_object should call do_something()')\n",
    "\n",
    "    # Utility functions\n",
    "    def setUp(self):\n",
    "        try:\n",
    "            self.assertRaisesRegex\n",
    "        except AttributeError:\n",
    "            self.assertRaisesRegex = self.assertRaisesRegexp\n",
    "\n",
    "    def assertRaisesWithMessage(self, exception):\n",
    "        return self.assertRaisesRegex(exception, r\".+\")\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], exit=False)\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 2
}
